tests/tensor/test_math.py

```diff
@@ -760,9 +761,11 @@ def test_isnan():
 
 class TestMaxAndArgmax:
     def setup_method(self):
-        MaxAndArgmax.debug = 0
+        Max.debug = 0
+        Argmax.debug = 0
 
     def test_basic(self):
+        # dbt: for some reason, Argmax does not work when I pass: n = as_tensor_variable(5.0)
```

Comment:
Scalars are still a problem. Am working on this and looking for how numpy handles them as suggested.

Reply:
It's raising from this C-code check: https://github.com/pymc-devs/pytensor/blob/8e7f626c6aebc2de75b90ab37b694958d4f68302/pytensor/tensor/math.py#L227

It seems not to handle the empty axes case that's passed when it's a scalar. Instead of empty axes we can convert to `None`.

But actually this function seems to be creating the problem in the first place? https://github.com/pymc-devs/pytensor/blob/8e7f626c6aebc2de75b90ab37b694958d4f68302/pytensor/tensor/math.py#L403

I don't think it should be needed or this convoluted. We handle axes in other places with way less code. We should use the numpy helper like we do for other cases, or let axes = None, which both argmax and max support anyway.

Reply:
I was not seeing problems before because I was creating `Argmax(axis=None)(pt.as_tensor(5.0))` directly. Which shows the problem is how the helper is creating the Argmax, basically `Argmax(axis=())`. Actually I am not sure what `Argmax(axis=())` should do, I think it should return `zeros_like(x)` since it corresponds to a no-reduction. `np.max(x, axis=()` just returns x as well. We should check our Max does the same btw, which I think it's not doing.

Reply:
I tried the 

> It's raising from this C-code check:
> 
> https://github.com/pymc-devs/pytensor/blob/8e7f626c6aebc2de75b90ab37b694958d4f68302/pytensor/tensor/math.py#L227
> 
> It seems not to handle the empty axes case that's passed when it's a scalar. Instead of empty axes we can convert to `None` or zero which is equivalent for the scalar. I prefer `None` because that's the default anyway.
> 
> But actually these lines seem to be creating the problem in the first place?
> 
> https://github.com/pymc-devs/pytensor/blob/8e7f626c6aebc2de75b90ab37b694958d4f68302/pytensor/tensor/math.py#L403-L405
> 
> I don't think it should be needed or this convoluted (referring to that `check_and_normalize_axes`, which I think is only used here?). We handle axes in other places with way less code. We should use the numpy helper like we do for other cases, or let axes = None alone, which both argmax and max support anyway.

I tried this locally and it works with a silly modification:
```
def max_and_argmax(a, axis=None, keepdims=False):
    """
    Returns maximum elements and their indices obtained by iterating over
    given axis.

    When axis is None (the default value), the max is performed
    over the flattened tensor.

    Parameters
    ----------
    keepdims : bool
        If this is set to True, the axes which are reduced are left in
        the result as dimensions with size one. With this option, the result
        will broadcast correctly against the original tensor.

    """
    # Check axis and convert it to a Python list of integers.
    # Axis will be used as an op param of MaxAndArgmax.
    a = as_tensor_variable(a)
    axis = check_and_normalize_axes(a, axis)
    if len(axis) == 0:
        axis = None
    out = Max(axis)(a)
    argout = Argmax(axis)(a)

    if keepdims:
        out = makeKeepDims(a, out, axis)
        argout = makeKeepDims(a, argout, axis)
    return [out, argout]
```
Scalars work this way. But now in the ```grad``` function for max, the line ```axis = as_tensor_variable(self.axis)``` misbehaves as self.axis is ```None``` and as_tensor_variable(None) is wrong. Maybe doing ```if self.axis is None:
            self.axis= list(range(x.ndim))``` help but will it be correct?

Reply:
> I was not seeing problems before because I was creating `Argmax(axis=None)(pt.as_tensor(5.0))` directly. Which shows the problem is how the helper is creating the Argmax, basically `Argmax(axis=())`. Actually I am not sure what `Argmax(axis=())` should do, I think it should return `zeros_like(x)` since it corresponds to a no-reduction. `np.max(x, axis=())` just returns x as well. We should check our Max does the same btw, which I think it's not doing.

Do the current changes reflect on this? The case of ```pt.as_tensor(5.0)``` is handled effectively now ig.
And the assert https://github.com/pymc-devs/pytensor/blob/e48ff56010f42ef4e656ddbfe253b4ffc93f500c/tests/tensor/test_math.py#L768
does not give any error so I assume it is doing what we expect it to do?

Reply:
Even 
```
n = as_tensor_variable(5.0)
        v, i = eval_outputs(max_and_argmax(n, axis=()))
        assert v == 5.0
        assert i == 0
        assert i.dtype == "int64"
        v = eval_outputs(max_and_argmax(n)[0].shape)
        assert len(v) == 0
        v = eval_outputs(max_and_argmax(n)[1].shape)
        assert len(v) == 0
```
works fine so I assume ```axis=()``` works as expected?